File: TimeSeriesSRC/basefunctions/uniAnal.py
import numpy as np
import matplotlib.pyplot as plt
from .sdiff import func_sdiff
from .xcorr import func_xcorr
from .plotgpac import func_plotgpac
from .parcor import func_parcor
from .gpac import func_gpac
from .makerow import func_makerow

def func_uniAnal (y,na=20,nump=10,nrg=5,ncg=0,diff=[0],per=[],perdsp=1) :
	"""Compute and plot the ACF, PACF, and GPAC for a univariate time series.

    Differences the series as requested, computes the autocorrelation (ACF),
    partial autocorrelation (PACF), and generalized partial autocorrelation
    (GPAC) functions, and produces stem/GPAC plots for order identification.

    Parameters
    ----------
    y : array-like
        1-D time series.
    na : int, optional
        Maximum lag for ACF; lags range from ``-na`` to ``+na``. Default 20.
    nump : int, optional
        Number of PACF terms to compute (lags 1 .. nump). Default 10.
    nrg : int, optional
        Number of GPAC rows (numerator orders). Default 5.
    ncg : int, optional
        Number of GPAC columns (denominator orders); 0 sets it equal to
        ``nrg``. Default 0.
    diff : list of int, optional
        Differencing orders to apply before analysis.
        Must satisfy ``len(diff) == len(per) + 1``. Default ``[0]``.
    per : list of int, optional
        Seasonal periods.  Used with ``diff`` — must have one fewer element.
        Default ``[]``.
    perdsp : int, optional
        Display period — ACF and PACF are sampled at every ``perdsp``-th lag.
        Default 1.

    Returns
    -------
    yacf : ndarray, shape (1, 2*na+1)
        Biased ACF from lag ``-na`` to ``+na``.
    ypacf : ndarray, shape (1, nump)
        Partial ACF from lag 1 to ``nump``.
    ygpac : ndarray, shape (nrg, ncg)
        GPAC table (row = numerator order, column = denominator order).

    Examples
    --------
    >>> import numpy as np
    >>> from scipy.signal import lfilter
    >>> from TimeSeriesSRC.basefunctions.uniAnal import func_uniAnal
    >>> e = np.random.default_rng(0).standard_normal(500)
    >>> y = lfilter([1], [1, -0.8], e)
    >>> yacf, ypacf, ygpac = func_uniAnal(y, na=20, nump=10, nrg=5)

    See Also
    --------
    multiAnal : Analysis for input-output (transfer function) data.
    uniChi : Box-Pierce chi-square whiteness test on residuals.
	"""


	y = func_makerow(y)

	if ncg == 0:
		ncg = nrg

	numpts = len(y[0])

	if (len(per) + 1) != len(diff):
		error='per must have one less term than diff.'
		raise Exception(error)

	# Difference the sequence

	period = np.append([1],diff)

	for i in range(len(diff)):

		d = diff[i]
		if (d!=0):
			y = func_sdiff(y, d, period[i]);

	tot = max([nump + 1, nrg + ncg])

	if na < tot:
		L = tot * perdsp
	else:
		L = na * perdsp

	if L > numpts:
		error='Not enough data to compute the acf for sufficient lags.'
		raise Exception(error)


	yacf = func_xcorr(y, y, L, 'biased');

	# Select data only at multiples of the display period

	xlen = len(yacf[0])
	if perdsp > 1:
		m =yacf[0,(L - perdsp):: -perdsp]
		n =yacf[0,L :xlen:perdsp]
		yacf =  np.append(m[::-1],n)
		yacf = yacf.reshape(1,-1)
		# yacf = [fliplr(yacf((L + 1 - perdsp):-perdsp: 1)) yacf((L + 1): perdsp:xlen)];


	# Calculate pacf and gpac

	ypacf, phi, sigma = func_parcor(yacf[0], nump);
	ygpac = func_gpac(yacf, nrg, ncg);


	# Convert acf to proper length for plotting

	xlen = len(yacf[0]);

	L = round((xlen - 1) / 2);


	if L != na:
		dif = L - na
		yacf = yacf[0,dif:xlen - dif]

	lag = np.arange(-na,na+1)

	confint = yacf[0,na] * 2 / np.sqrt(numpts)

	# Getfigure if it exists, or create new figure

	me = 'ACF and PACF'

	fig, (ax1, ax2) = plt.subplots(2)
	fig.suptitle(me)
	ax1.stem(lag, yacf[0])


	ax1.set_title('Autocorrelation Function')
	ax1.set_xlabel('Lag')
	xlim = ax1.get_xlim()
	ax1.plot([xlim[0], xlim[1]], [0 ,0], 'k');
	ax1.plot([xlim[0], xlim[1]], [ confint, confint], ':r');
	ax1.plot([xlim[0], xlim[1]], [-confint,-confint], ':r');

	ind = [i for i in range(1,nump+1)]
	ax2.stem(ind,ypacf[0])
	ax2.set_title('Partial Autocorrelation Function')
	ax2.set_xlabel('Lag')
	xlim = ax2.get_xlim()
	ax2.plot([xlim[0], xlim[1]], [0 ,0], 'k');

	# Get figure if it exists, or create new figure

	plt.tight_layout()
	plt.show()

	func_plotgpac(ygpac, 'GPAC Array')

	return yacf, ypacf, ygpac

File: Matlab_Files/PredictMod/uniAnal.m
function [yacf,ypacf,ygpac] = uniAnal(y,na,np,nrg,ncg,diff,per,perdsp)
%UNIANAL Univariate analysis
%	
%	Syntax
%
%	  [yacf,ypacf,ygpac] = uniAnal(y,na,np,nrg,ncg,diff,per,perdsp)
%
%	Description
%	
%	  UNIANAL computes the autocorrelation and partial
%	  autocorrelation functions and the generalized partial
%	  autocorrelation array of the sequence y and
%	  plots the results.
%	
%	  UNIANAL(Y,NA,NP,NRG,NCG,DIFF,PER,PERDSP) takes these inputs,
%	    Y   - Vector containing the sequence.
%	    NA  - The autocorrelation lags will range from -NA to +NA; default = 20.
%	    NP  - Number of partial autocorrelation terms to compute; default = 10.
%	    NRG - Number of rows of the GPAC to compute; default = 5.
%	    NCG - Number of columns of the GPAC to compute; default = 5.
%	    DIFF  = [diff1 diff2...diffNP], default = [0].
%	      diffi - Order of the differencing for period i.
%	    PER   = [per1 per2...perNP], default = [].
%	      peri  - Period i.
%	    PERDSP- Period at which to display acf and pacf data.; default = 1;
%	  and returns,
%	    YACF  - Autocorrelation function for Y.
%	    YPACF - Partial autocorrelation function for Y.
%	    YGPAC - Generalized partial autocorrelation function for Y.
%		
%	Examples
%
%	  This code generates an autoregressive sequence.
%	
%	    e = randn(1,2000);
%	    y = filter(1,[1 -.8],e);
%
%	  The following command generates the autocorrelation and
%	  partial autocorrelation functions.  The acf will be
%	  computed from lag -20 to lag 20.  The pacf will be computed
%	  from order 1 to order 10. 
%
%	    [yacf,ypacf,ygpac] = uniAnal(y,20,10);
%

% Yong Hu, Martin Hagan, 9-15-00
% $Revision: 1.0 $ $Date: 21-Sep-2000 14:37:36 $

if nargin<2, na = 20;end
if nargin<3, np = 10;end
if nargin<4, nrg = 5;end
if nargin<5, ncg = nrg;end
if nargin<6, diff = 0;end
if nargin<7, per = [];end
if nargin<8, perdsp = 1;end

numpts = length(y);

if (length(per)+1) ~= length(diff),
  error('per must have one less term than diff.')
end

% Difference the sequence
period = [1 per];
for i=1:length(diff),
  d = diff(i);
  if (d~=0),
    y = sdiff(y,d,period(i));
  end
end

tot = max([np+1 nrg+ncg]);
if na<tot,
  L = tot*perdsp;
else
  L = na*perdsp;
end

if L>numpts,
  error('Not enough data to compute the acf for sufficient lags.');
end

yacf = xcorr(y,y,L,'unbiased');
len = length(yacf);

% Select data only at multiples of the display period
if perdsp>1,
  yacf = [fliplr(yacf((L+1-perdsp):-perdsp:1)) yacf((L+1):perdsp:len)];
end

% Calculate pacf and gpac
ypacf = parcor(yacf,np);
[ygpac]=gpac(yacf,nrg,ncg);

% Convert acf to proper length for plotting
len = length(yacf);
L = round((len-1)/2);
if L~=na,
  dif = L-na;
  yacf = yacf(1+dif:len-dif);
end

lag = -na:na;

confint = yacf(na+1)*2/sqrt(numpts);

% Get figure if it exists, or create new figure
me = 'ACF and PACF';
fig1 = chkfig(me);
if length(get(fig1,'children')) == 0, fig = 0; end
if (exist('fig1')&&(fig1~=0))
  figure(fig1);
else
  fig1 = figure;
  set(fig1,'name',me,'numbertitle','off')
end

subplot(2,1,1), bar(lag,yacf,.4)
title('Autocorrelation Function')
xlabel('Lag')
hold on
ax = axis;
plot([ax(1) ax(2)],[0 0],'k');
plot([ax(1) ax(2)],[confint confint],':r');
plot([ax(1) ax(2)],[-confint -confint],':r');
hold off
subplot(2,1,2), bar(ypacf,.4)
title('Partial Autocorrelation Function')
xlabel('Lag')
hold on
ax = axis;
plot([ax(1) ax(2)],[0 0],'k');
hold off

% Get figure if it exists, or create new figure
me = 'GPAC Array';
fig2 = chkfig(me);
if length(get(fig2,'children')) == 0, fig = 0; end
if (exist('fig2')&&(fig2~=0))
  figure(fig2);
else
  fig2 = figure;
  set(fig2,'name',me,'numbertitle','off')
end

plotgpac(ygpac,'GPAC Array');
